### ISSUE TITLE
GPU yields worse metric than CPU

### ISSUE THREAD

**Issue by kdlin** [catboost/catboost#1257]
Problem:various measurements become worse when I switch from CPU to GPU
catboost version:0.22
Operating System:Linux 4.4.0-1100-aws x86_64
CPU: Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz
# GPU: Tesla M60

I wanted to reduce the training time and so I specified 'task_type' as 'GPU'. I immediately noticed that its metrics got worse. The only change I made was setting task_type as GPU. The rest are the same. 

The training dataset has 1.2M rows and 218 columns. Among these 218 columns, 42 are categorical features. The rest are floats or integers, no text features. The validation dataset has 120K rows.

The following are the parameters for the CPU version:
{'nan_mode': 'Min', 'eval_metric': 'Logloss', 'combinations_ctr': ['Borders:CtrBorderCount=15:CtrBorderType=Uniform:TargetBorderCount=1:TargetBorderType=MinEntropy:Prior=0/1:Prior=0.5/1:Prior=1/1', 'Counter:CtrBorderCount=15:CtrBorderType=Uniform:Prior=0/1'], 'iterations': 1000, 'sampling_frequency': 'PerTree', 'fold_permutation_block': 0, 'leaf_estimation_method': 'Newton', 'od_pval': 0, 'counter_calc_method': 'SkipTest', 'grow_policy': 'SymmetricTree', 'boosting_type': 'Plain', 'model_shrink_mode': 'Constant', 'feature_border_type': 'GreedyLogSum', 'ctr_leaf_count_limit': 18446744073709551615, 'bayesian_matrix_reg': 0.10000000149011612, 'one_hot_max_size': 2, 'l2_leaf_reg': 3, 'random_strength': 1, 'od_type': 'Iter', 'rsm': 1, 'boost_from_average': False, 'max_ctr_complexity': 4, 'model_size_reg': 0.5, 'simple_ctr': ['Borders:CtrBorderCount=15:CtrBorderType=Uniform:TargetBorderCount=1:TargetBorderType=MinEntropy:Prior=0/1:Prior=0.5/1:Prior=1/1', 'Counter:CtrBorderCount=15:CtrBorderType=Uniform:Prior=0/1'], 'subsample': 0.800000011920929, 'use_best_model': True, 'od_wait': 35, 'class_names': [0, 1], 'random_seed': 42, 'depth': 6, 'ctr_target_border_count': 1, 'has_time': False, 'store_all_simple_ctr': False, 'border_count': 254, 'classes_count': 0, 'sparse_features_conflict_fraction': 0, 'leaf_estimation_backtracking': 'AnyImprovement', 'best_model_min_trees': 1, 'model_shrink_rate': 0, 'min_data_in_leaf': 1, 'loss_function': 'Logloss', 'learning_rate': 0.30000001192092896, 'score_function': 'Cosine', 'task_type': 'CPU', 'leaf_estimation_iterations': 10, 'bootstrap_type': 'MVS', 'max_leaves': 64, 'permutation_count': 4}

The following are the parameters for the GPU version:
{'nan_mode': 'Min', 'gpu_ram_part': 0.95, 'eval_metric': 'Logloss', 'combinations_ctr': ['Borders:CtrBorderCount=15:CtrBorderType=Uniform:TargetBorderCount=1:TargetBorderType=MinEntropy:Prior=0/1:Prior=0.5/1:Prior=1/1', 'FeatureFreq:CtrBorderCount=15:CtrBorderType=Median:Prior=0/1'], 'iterations': 1000, 'fold_permutation_block': 64, 'leaf_estimation_method': 'Newton', 'observations_to_bootstrap': 'TestOnly', 'od_pval': 0, 'counter_calc_method': 'SkipTest', 'grow_policy': 'SymmetricTree', 'boosting_type': 'Plain', 'ctr_history_unit': 'Sample', 'feature_border_type': 'GreedyLogSum', 'bayesian_matrix_reg': 0.10000000149011612, 'one_hot_max_size': 2, 'devices': '-1', 'pinned_memory_bytes': '104857600', 'l2_leaf_reg': 3, 'random_strength': 1, 'od_type': 'Iter', 'rsm': 1, 'boost_from_average': False, 'fold_size_loss_normalization': False, 'max_ctr_complexity': 4, 'gpu_cat_features_storage': 'GpuRam', 'simple_ctr': ['Borders:CtrBorderCount=15:CtrBorderType=Uniform:TargetBorderCount=1:TargetBorderType=MinEntropy:Prior=0/1:Prior=0.5/1:Prior=1/1', 'FeatureFreq:CtrBorderCount=15:CtrBorderType=MinEntropy:Prior=0/1'], 'use_best_model': True, 'od_wait': 35, 'class_names': [0, 1], 'random_seed': 42, 'depth': 6, 'ctr_target_border_count': 1, 'has_time': False, 'border_count': 128, 'min_fold_size': 100, 'data_partition': 'FeatureParallel', 'bagging_temperature': 1, 'classes_count': 0, 'leaf_estimation_backtracking': 'AnyImprovement', 'best_model_min_trees': 1, 'min_data_in_leaf': 1, 'add_ridge_penalty_to_loss_function': False, 'loss_function': 'Logloss', 'learning_rate': 0.30000001192092896, 'score_function': 'Cosine', 'task_type': 'GPU', 'leaf_estimation_iterations': 10, 'bootstrap_type': 'Bayesian', 'max_leaves': 64, 'permutation_count': 4}

**Comment by kdlin** [comment]
Also, the GPU version is about 3.5X-4.2X faster in training. According to the user guide, for large datasets (> 1M rows), one should see 40X-50X speed improvement. But I get only 1/10 of that. How do I improve the GPU speed?

**Comment by kdlin** [comment]
Found a similar bug from the past: https://github.com/catboost/catboost/issues/241

**Comment by pyanishjain** [comment]
yes you are write i face the similar issues with this as well as with xgboost algo

**Comment by annaveronika** [comment]
Here is the explanation of differences between CPU and GPU trainings:
https://catboost.ai/docs/concepts/faq.html#differencies-between-cpu-and-gpu

You can set the parameters that are defferent between CPU and GPU to the same value, and you will get the same quality.

Hope it helps!

**Comment by kdlin** [comment]
Can you please comment if it is possible to get the same value between CPU and GPU since their parameters are different? For example, the boostrap type for CPU is MVS and for GPU it is Bayesian. One cannot set them the same due to the limitation of the CatBoost implementation. Is it realistic to expect that they can yield the same value?

**Comment by annaveronika** [comment]
There will not be exactly the same value, also if you run GPU training for several times, quality will be a little different. But if you set all different parameters to the same values, then quality will be very similar between CPU and GPU.
In some cases it is not possible to get the same value, this happens if some functionality is supported only on CPU or only on GPU. For example, MVS sampling is only supported on CPU, so you can use Bayesian sampling on CPU and GPU to get the same results.
MVS sampling will be implemented on GPU in one of the next releases.

**Comment by kdlin** [comment]
Thanks for the explanation. I have found CatBoost yields better "quality" (I think this is your term for better performance metrics.) than LightGBM and XGBoost for the majority of datasets I am working on. As you know, CatBoost runs much slower in CPU mode, it runs faster in the GPU mode (provided the GPU has sufficient built-in RAM), but GPU always yields worse quality. As a net result, CatBoost is just marginally better than LightGBM and XGBoost. It will be very desirable if the GPU mode can yield the same level of quality as the CPU mode. Thanks again for the response.

**Comment by kizill** [comment]
@kdlin speaking about quality, what difference did you notice? And are you sure you are not comparing quality of underfitted models?

**Comment by kdlin** [comment]
@kizill I have millions of rows of data to train the model. I also use hyeropt to tune the hyperparameters. So I don't think these models are underfitted. A good measurement of imbalanced dataset is Precision Recall AUC (as opposed to ROC-AUC that is commonly used.) This is the most difficult metric to improve. I have found CatBoost in CPU mode yields better PR-AUC by 2-3% than other gradient boosting algorithms. However, in GPU mode, CatBoost loses about 1+% PR-AUC.

**Comment by kdlin** [comment]
@kizill and @annaveronika I'd like to request that Precision Recall ROC be added as one of the built-in metrics for binary classification. XGBoost has it. See this link and search for 'aucpr'. https://xgboost.readthedocs.io/en/latest/parameter.html

**Comment by kizill** [comment]
@kdlin speaking about PR AUC - we are currently reviewing pull request https://github.com/catboost/catboost/pull/1249 now, so it'll be shipped in some next releases 🙂
And thank you a lot for giving us numbers!

**Comment by kdlin** [comment]
@kizill You are welcome. Keep up the good work. I have found your handling of categorical features is  very useful.

**Comment by kizill** [comment]
Speaking about 1% quality drop - GPU and CPU implementation differs so much so it's impossible to say whether it is a bug or implementation detail. For example, not only GPU and CPU have different subset of implemented features - we use double precision floats to accumulate histogram values on CPU and single precision on gpu. But we consider such a large drop as strange and will try to investigate it. Is your dataset publicly available?

**Comment by kdlin** [comment]
It is 1+% drop. This makes it not attractive to use CatBoost because: (1) GPU AWS intances are much more expensive, (2) the improvement above XGBoost and LightGBM is marginal with GPU, and (3) CatBoost in CPU takes too long to tune parameters. I hope you can address this issue. I think you guys are onto something. Sorry the dataset is not publically available.

**Comment by kdlin** [comment]
Note that 1% means the difference between PR-AUC scores of 0.55 and 0.56. A 3% boost is like from PR-AUC score of 0.55 to 0.58. This is quite a substantial improvement for imbalanced datasets!

**Comment by annaveronika** [comment]
You should try setting exactly the same parameters for CPU and GPU trainings and check resulting metrics

**Comment by annaveronika** [comment]
You could also provide your dataset, so that we could check the same.

**Comment by annaveronika** [comment]
I suppose that the difference that you are seeing is caused by the difference in the parameters: if parameters are different, then difference in metrics is ok.
To check that you need to set parameters to the same value, it's always possible. If something is not supported on GPU, you need to set this parameter on CPU to the value is supported on GPU. And vice versa.
Also some metric changes might be caused by randomness and floating point summation order, to check that you need to run gpu training several times, preferably with different random seed values.

**Comment by NasirKhalid24** [comment]
I'm currently facing the same issue and so far have moved as many parameters as possible to match the CPU settings:

```
border_count=254,
bootstrap_type="MVS",
boosting_type="Plain",
```

My AUC between training and test over 5 folds - 3 repeats changes by a factor of ~10. Are there are settings I've forgotten to change?

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
In order to find out which parameters differ, you may call get_all_params() for the CPU and GPU boosters.

**Comment by NasirKhalid24** [comment]
> In order to find out which parameters differ, you may call get_all_params() for the CPU and GPU boosters.

Thanks! I tried this and discovered that the main reason is in the Ctr-type, the Counter in CPU is making all the difference

**Comment by RiverShah** [comment]
Requesting that this issue should be investigated more please. It is simply not a matter of matching parameters (which I have done so as much as possible between CPU vs GPU). Please try with many synthetic datasets with imbalanced classes and low signal to noise ratio. GPU's quality is significantly worse than CPU baselines. 

@kizill You mentioned single vs double precision for histogram formation for CPU vs GPU. My suspicion is that this can easily be causing the observed learning issues. Is it possible to generate the histogram using double precision on the GPU please? I am using Tesla V100s. Will double precision histogram give up a significant portion of the GPU speedups?
